cholla-hydro/cholla
# Add testing to Cholla (in beta)

# General Testing Notes:
- All new functions, classes, etc should have doxygen documentation in header
  files or next to the definition if there is no header file for it.
- Comparisons of floats can't be done with gtest's `ASSERT_DOUBLE_EQ` since it
    only performs a ULP check and some of the small differences in results
    between compilers have large ULP distances between them. Just comparing the
    results of compiling Cholla with GCC vs XL gives many differences greater
    than 4 ULP, sometimes much greater, and ASSERT_DOUBLE_EQ only allows 4 ULP.
    I added a custom equality check instead in the `testingUtilities` namespace.
- Added a new `testingUtilities` namespace in
  `cholla/src/utils/testingUtilities.h` to hold various functions, classes, etc
  required for testing
- Added the new `testingUtilities::ulpsDistanceDbl` function to compute the ULP
  distance between two double precisions (FP64) floating point numbers. It also
  performs various edge case checks to avoid  returning nonsensical or erroneous
  results
- Added the `testingUtilities::nearlyEqualDbl` function to check if two FP64
  numbers are "close enough" where "close enough" is an order of magnitude
  larger than the difference in results between the XL and GCC compilers or that
  the difference between them is less than 4 ULP. Both the absolute margin and
  ULP margin can be set by the user but default to reasonable values detailed in
  the documentation for `testingUtilities::nearlyEqualDbl`
- Versions of both `testingUtilities::nearlyEqualDbl` and
  `testingUtilities::ulpsDistanceDbl` that support floating point precisions
  other than FP64 are easily implemented with templates but are not required for
  the testing beta
- Added a custom main function for tests rather than using Googletest's default
  main function
  - This is required to gather command line arguments for where the cholla
    executable is, the build type, and the machine being used, etc. These are
    all stored as global variables and instances of the
    `testingUtilities::GlobalString` class
  - Added the `testingUtilities::GlobalString` class for storing a single
    string that's read only (convention not compiler enforced).
  - Added the `globalChollaRoot` global variable to hold the path to the
    root of Cholla
  - Add `globalChollaBuild` and `globalChollaMachine` as global variables
  - Add argument parsing to `main_tests`
- Added a script to run tests in `cholla/builds/scripts/run_tests.sh`. Details
  on how to use it can be found in the testing page of the Cholla wiki
- Tests don't currently build on Summit when using GCC, XL works fine. PR #93 
  might fix this

# HLLC Test:
- A simple test for the HLLC CUDA solver that uses the `nearlyEqualDbl` function

# Added A Class for system tests:
- Added a new class for system tests (`SystemTestRunner`), a namespace to hold
  it (`systemTest`). Details on usage can be found in the Testing page on the
  Cholla wiki.
- The `SystemTestRunner` class supports MPI runs of Cholla, any runs with a
  large number of ranks, more than a few tens, should consider using the command
  line flags provided to run Cholla in a large job allocation then do the
  comparisons in a job with a single node as the comparisons are not performed
  in parallel. Parallel comparisons are not done since they're very fast and
  even a 50k rank job where each GPU has a 10^3 grid should take about a minute
  to do the comparison on a single thread.
- Added a basic hydro Sod Shock Tube system test
  - 1, 2, and 4 rank version of this test were added. The 4 rank version uses a
    larger grid (64^3 instead of 16^3) to avoid issue #92.
- Limitations:
  - The 4 rank Sod test fails on Spock, likely due to issue #92
  - Particles are not supported since I couldn't get particle runs to produce
    HDF5 output without throwing lots of errors. This should be fixed with PR
    #90 and once that's integrated then adding support for particles should be
    straightforward

# Build System Integration:
- Rather than create a new build type for testing integrating testing into the
  build system was done largely by filtering the source files. If the `TEST`
  makefile variable is not set to `true` then building Cholla commences as
  normal, the only change is that all files that end in `_tests.extension` are
  filtered out of the source list. If `TEST` is set to `true` then Cholla's
  usual `main` function is filtered out, Googletest and the HDF5 C++ API are
  added to the linking list, and some output is provided to the user to indicate
  they're building with tests.
- This filtering does mean the `OBJS` variable does not always contain a
  complete list of potential object files to use when running `make clean`. To
  fix this an unfiltered `CLEAN_OBJS` variable is genated before any filtering
  takes place and is used in the `make clean` target
- Tests executables now end in `.tests` and are properly cleaned when using
  `make clean` or `make clobber`
- Automatically set gtest root on different machines in the machine specific
  Makefiles
- `make clobber` now removes XML testing reports in `cholla/bin/` and all
  directories in `cholla/bin/` that start with `t` in an effort to delete the
  directories that the system tests generate.

# Automated Testing
Automated testing of Cholla requires hardware that is not common amongst CI
providers. The University of Pittsburgh Center for Research Computing is looking
into providing their cluster resources as a place that CI pipelines can be run
and so automated testing with Cholla will wait until they've finished. In the
meantime, I've added a GitHub Actions workflow matrix job that builds Cholla and
the tests with all make types and against both CUDA and HIP. Note that it
doesn't run the tests or Cholla, just checks that they build. CUDA, ROCm, Clang,
and GCC versions could all easily be added to the job matrix, it would just
require some manual filtering to remove unnecessary jobs.
- Uses ROCm 4.3.1 as the new 4.5 doesn't install properly on GitHub hosted
  runners
- Added a readme badge to indicate status of build

# Other:
- Git ignore HDF5 files with *.h5.* pattern
- Remove all files related to old regression test.
- Rename the `tests` directory to `examples` to avoid future confusion.
- Fix all mentions of the old `tests` directory, mentions in the wiki will be
  fixed when the testing documentation is added.
- Build support for C-3PO
